src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -57,6 +56,106 @@
 #include "utilities/events.hpp"
 #include "utilities/preserveException.hpp"
 
+void MonitorList::add(ObjectMonitor* m) {
+  for (;;) {
+    ObjectMonitor* head = Atomic::load(&_head);
+    m->set_next_om(head);
+    if (Atomic::cmpxchg(&_head, head, m) == head) {
+      break;
+    }
+  }
+  size_t count = Atomic::add(&_count, 1u);
+  if (count > max()) {
+    Atomic::inc(&_max);
+  }
+}
+
+size_t MonitorList::count() const {
+  return Atomic::load(&_count);
+}
+
+size_t MonitorList::max() const {
+  return Atomic::load(&_max);
+}
+
+// Walk the in-use list and unlink (at most MonitorDeflationMax) deflated
+// ObjectMonitors. Returns the number of unlinked ObjectMonitors.
+size_t MonitorList::unlink_deflated(Thread* self, LogStream* ls,
+                                    elapsedTimer* timer_p,
+                                    GrowableArray<ObjectMonitor*>* unlinked_list) {
+  size_t unlinked_count = 0;
+  ObjectMonitor* prev = NULL;
+  ObjectMonitor* head = Atomic::load_acquire(&_head);
+  ObjectMonitor* m = head;
+  while (m != NULL) {
+    if (m->is_being_async_deflated()) {
+      // Find next live ObjectMonitor.
+      ObjectMonitor* next = m;
+      while (next != NULL && next->is_being_async_deflated()) {
```

Comment:
Nit: This loop seems odd. Given we know m is_being_async_deflated, this should either be a do/while loop, or else we should initialize:

ObjectMonitor* next = m->next_om();

and dispense with the awkwardly named next_next.

Reply:
@fisk - I'm leaving this one for you for now.

Reply:
Changing it to a do/while loop makes sense. The while condition is always true the first iteration, so doing a while or do/while loop is equivalent. If you find the do/while loop easier to read, then that sounds good to me.

Reply:
Okay. I've changed it to a do-while loop.

Reply:
The loop that was being discussed here is the one on line 93 of the original changeset:
`while (next != NULL && next->is_being_async_deflated()) {`
I made no comment on the outer while loop.

Reply:
You're correct. I got confused by the line renumbering due to other changes.
Fortunately, the outer while loop can also correctly and easily be changed
into a do-while loop so I'll keep that change. I'm looking at the correct inner
while loop now.

Reply:
The inner while loop is now converted into a do-while loop.